presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1560,12 +1537,19 @@ private void analyzeWith(Query node, AnalysisContext context)
                     throw new SemanticException(MISMATCHED_COLUMN_ALIASES, withQuery, "WITH column alias list has %s entries but WITH query(%s) has %s columns", columnNames.size(), name, queryDescriptor.getVisibleFieldCount());
                 }
             }
+            if (withScopeBuilder.containsNamedQuery(name)) {
```

Comment:
I am not sure if understand question correctly. Anyway I will try to answer it :)

`Scope.Builder` does have a knowledge about the AST nodes, throwing `SemanticException` from it would be strange. So this is why I have to check if query is not already added to builder before adding it there.